src/libstore/gcs-binary-cache-store.cc

```diff
@@ -0,0 +1,194 @@
+#if ENABLE_GCS
+
+#include "nar-info.hh"
+#include "nar-info-disk-cache.hh"
+#include "globals.hh"
+#include "compression.hh"
+#include "binary-cache-store.hh"
+
+#include <memory>
+#include <google/cloud/storage/client.h>
+
+namespace gcs = google::cloud::storage;
+
+using namespace std::chrono_literals;
+using ::google::cloud::StatusOr;
+
+namespace nix {
+
+struct GCSBinaryCacheStore : public BinaryCacheStore
+{
+    const Setting<std::string> narinfoCompression{this, "", "narinfo-compression", "compression method for .narinfo files"};
+    const Setting<std::string> lsCompression{this, "", "ls-compression", "compression method for .ls files"};
+    const Setting<std::string> logCompression{this, "", "log-compression", "compression method for log/* files"};
+    const Setting<uint64_t> bufferSize{
+        this, 5 * 1024 * 1024, "buffer-size", "size (in bytes) of each downloaded chunk"};
+
+    std::string bucketName;
+    std::unique_ptr<gcs::Client> client;
+
+    GCSBinaryCacheStore(
+            const Params & params, const std::string & bucketName)
+        : BinaryCacheStore(params)
+        , bucketName(bucketName)
+        , client(nullptr)
+    {
+        diskCache = getNarInfoDiskCache();
+    }
+
+    std::string getUri() override
+    {
+        return "gs://" + bucketName;
+    }
+
+    void init() override
+    {
+        if (client == nullptr) {
+            StatusOr<gcs::ClientOptions> options =
+                gcs::ClientOptions::CreateDefaultClientOptions();
```

Comment:
assuming we're not on a gcs-based machine that has its credentials provisioned by the metadata server…

can we just pass `GOOGLE_APPLICATION_CREDENTIALS`?

Reply:
Yes, I documented that in the docbook section for GCS.

Reply:
Right, somehow missed the docbook section. Thanks!